src/librustc_mir/dataflow/impls/indirect_mutation.rs

```diff
@@ -0,0 +1,148 @@
+use rustc::mir::visit::Visitor;
+use rustc::mir::{self, Local, Location};
+use rustc::ty::{self, TyCtxt};
+use rustc_data_structures::bit_set::BitSet;
+use syntax_pos::DUMMY_SP;
+
+use crate::dataflow::{self, GenKillSet};
+
+/// Whether a borrow to a `Local` has been created that could allow that `Local` to be mutated
+/// indirectly. This could either be a mutable reference (`&mut`) or a shared borrow if the type of
+/// that `Local` allows interior mutability.
+///
+/// If this returns `false` for a `Local` at a given `Location`, the user can assume that `Local`
+/// has not been mutated as a result of an indirect assignment (`*p = x`) or as a side-effect of a
```

Comment:
I should have said the latter, but yes "couldn't be so it hasn't been" XD. I think I'll add an example e.g. "If this returns `false` for a `Local` at the location of an indirect assignment, that `Local` *cannot* be mutated by that assignment. For example, if the dataflow state for a statement `*p = 42` contained locals `x` and `z` but not `y`, we know that while `x` or `z` may be the target of that assignment, `y` cannot be.

Reply:
See 1c4be91.

Reply:
Sorry, I actually mean "couldn't possibly have been mutated prior to the given `Location`". It's been awhile since I implemented this. This is accomplished because we never "kill" in this analysis, so as soon as a borrow exists that allows mutation, that `Local` is marked as indirectly mutable at all statements which that borrow can reach. I updated the docs to reflect this.